server/vcs/provider/github/check_runs_fetcher.go

```diff
@@ -0,0 +1,62 @@
+package github
+
+import (
+	"context"
+	"fmt"
+	gh "github.com/google/go-github/v45/github"
+	"github.com/palantir/go-githubapp/githubapp"
+	"github.com/pkg/errors"
+	"github.com/runatlantis/atlantis/server/events/models"
+	"net/http"
+	"regexp"
+)
+
+const (
+	CompletedStatus  = "completed"
+	FailedConclusion = "failed"
+)
+
+var checkRunRegex = regexp.MustCompile("atlantis/policy_check: .*")
+
+type CheckRunsFetcher struct {
+	ClientCreator githubapp.ClientCreator
+	AppID         int64
+}
+
+func (r *CheckRunsFetcher) ListFailedPolicyCheckRuns(ctx context.Context, installationToken int64, repo models.Repo, ref string) ([]string, error) {
+	client, err := r.ClientCreator.NewInstallationClient(installationToken)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating installation client")
+	}
+
+	var failedPolicyCheckRuns []string
+	nextPage := 0
+	for {
```

Comment:
in a past PR i asked if you could make this pagination generic, I think you made it work for the 2 usecases at that time, but now we're doing the same thing again here as done here:
https://github.com/lyft/atlantis/blob/7a0f8c9749907084f12c54ceaf662e994868b140/server/vcs/provider/github/file_fetcher.go#L40-L46

So I want to make sure that we can reuse code going forward. we are now on go 1.19 which supports generics, can we look into that for this usecase? Or if you don't want to do that figuring out a way to create a generic ListOptions paginator that returns some result we can use?